### ISSUE TITLE
[Project] Critical section for asynchrous work

### ISSUE THREAD

**Issue by tomaszkam** [atomgalaxy/review-executor-sendrecv#1]
The thing I would like to explore is writing something critical section for the asynchronous work, i.e. wrapper that would guarantee that specific part of works do not run in paller. To illustarte, if we have two pipelines/task:
```
async_mutex m;
auto work1 =
  start()
    .then(part1a)
    .then(locked(m, part1b))
    .then(part1c);

auto work2 =
  start()
    .then(part2a)
    .then(locked(m, part2b))
    .then(part2c);

thread_pool.execute(work1);
thread_pool.execute(work2);
// part1b and part2b does not run in parallel
```
The important aspect is that, in contrast to normal mutexes, this pools should not block execution agents. So the above code should work, even if we have only one thread (execution agent) in thread pool.

The major areas of investigation:
  1) Representing the asynchronous task using sender-receiver, I want part2 to not only be a simple function, but any work
  2) Avoiding executor switching - if work1 and work2 where started on different executor, then they should complete on the same
  3) Single allocation

**Comment by RobertLeahy** [comment]
My understanding is that this is a use case for the scheduler equivalent of a "strand" (an executor which doesn't allow submitted work to run in parallel).

I don't think the model would therefore be one of a mutex, rather it'd be a transition of scheduler. Consider the example in §1.4.2 of the paper and imagine `scheduler1` has the "strand" property whereas `scheduler2` does not.

**Comment by tomaszkam** [comment]
That's a bit different usage, I want the work to run on their respective execution context (not single schedule), but prevent from running at once.
The purpose of the experiment is the something as I done once - provide an understandable interface (mutex that prevents two threads running).

**Comment by tomaszkam** [comment]
I want to see if that model can be represented using sender/receiver.

**Comment by cor3ntin** [comment]
@tomaszkam I think you might be thinking of an async mutex, for example https://github.com/facebookexperimental/libunifex/blob/master/doc/api_reference.md#async_mutex 

It has the same interface as a mutex but its lock method returns a sender

```
co_await mutex.lock();
// critical section
```

**Comment by ericniebler** [comment]
Yes, it is possible to define a sender that implements an async mutex. cc @lewissbaker

**Comment by tomaszkam** [comment]
Starting with `capture_args` sender, that stores the values passed by the wrapped sender in the operation (as `std::variant`) and then passes a reference to it to the attached receiver.
Link: https://wandbox.org/permlink/5IPY23JR6sHOiHw4

**Comment by tomaszkam** [comment]
And `resume_via(schedule, args)` that allows to pass arguments from `capture_args` to receiver, and continue work on the specific `scheduler`.
Link: https://wandbox.org/permlink/xsTWEVewL77KXa5A

**Comment by tomaszkam** [comment]
Finally `locked(sender, work, mutex)` implementation, that runs work in a critical section (and unlock afterward).
Link:  https://wandbox.org/permlink/QBp9fBxTrdZEgSnz

**Comment by tomaszkam** [comment]
Moved the code to github, restored online compilation via godbolt:
https://godbolt.org/z/tKf37Q

**Comment by tomaszkam** [comment]
Added thread_pool and the test: https://godbolt.org/z/UtzRtR
Still missing proper joining of the thread_pool, with waiting for all task to be done.

**Comment by tomaszkam** [comment]
This is a very hacky implementation, that misses a lot of the proper concepts, reference propagating, noexcept, e.t.c.. My intent was to check that it is possible to implement the think I am looking for.

However, I have found a piece that I am missing - the ability to extract execution context, from a scheduler. The properties are working only with executors, and the only related is `contex_t`. 

I have added the `sender_with_scheduler` concept that exposes `scheduler` method on the sender. See: https://github.com/atomgalaxy/review-executor-sendrecv/commit/a8da1fc186a94c34e6c202ca3d5672d971e23451

**Comment by lewissbaker** [comment]
One approach to this is to create a scheduler-adapter that takes a scheduler and a mutex and produces a new scheduler that serialises execution of child operations.

e.g.
```
async_mutex mutex;
thread_pool_context tp;

// Here foo and bar will be serialised
sync_wait(when_all(locked(foo, mutex), locked(bar, mutex)));

// Construct a scheduler that will automatically serialise all work scheduled to it
// but still run on some underlying executor.
scheduler auto strand = make_strand_scheduler(tp.get_scheduler(), mutex);

// on(src, strand) is equivalent of locked(on(strand.innerScheduler), strand.mutex)

// Runs foo and bar serialised
sync_wait(when_all(on(foo, strand), on(bar, strand)));

// Different strands can use the same mutex but different schedulers.
scheduler auto strand1 = make_strand_scheduler(sched1, mutex);
scheduler auto strand2 = make_strand_scheduler(sched2, mutex); 

// Runs foo/bar serialised but on sched1/sched2 respectively.
sync_wait(when_all(on(foo, strand1), on(bar, strand2)));
```

It can also make `schedule()` hold the lock while calling `set_value()` and
implicitly call .unlock() when set_value() returns (or for lifetime safety
guarantees, possibly when the operation-state is destroyed).

**Comment by tomaszkam** [comment]
The strand like implementation, are a bit different from what I wanted to achieve. As I understand them they would be equivalent to performing all work needing a critical secrion from a single thread, and in this case scheduler.

My goal is to do not require changing od scheduler, and   instead provide quarantee that none of work will run concurrently. This, allow the usages that are more similiar to the to synchronous mutex, when I guarantee that no more than one thread is running.

**Comment by lewissbaker** [comment]
> My goal is to do not require changing od scheduler, and instead provide quarantee that none of work will run concurrently

Yes, this is what I had intended with my snippet.
My apologies if it I added confusion by using the name strand if that wasn't appropriate.

You can have many different scheduler instances, each of which would be a strand and would execute work on the underlying scheduler that it wraps. But each of these strands can share the same serialisation queue (ie. async_mutex) and so provide mutual exclusion across multiple execution contexts while preserving their respective execution contexts.

We have recently added something similar to folly to support this use-case.
See [folly::StrandExecutor](https://github.com/facebook/folly/blob/master/folly/executors/StrandExecutor.h) although note that this is not currently built on top of the scheduler concepts.

My point was more to consider as an alternative interface to (or perhaps in addition to) the `locked()` API for scheduling work with mutual exclusion, exposing this as a `scheduler` that adapts an underlying scheduler. Such an interface would interoperate better with other facilities defined in terms of `schedulers` and `on()`.

**Comment by tomaszkam** [comment]
Firstly, just want to note that this was only a toy project. I was considering the viability of such a specific interface.

Regarding the `locked_scheduler` how would I represent something like this (note it never mentions scheduler)?
```
auto s1 = work1();
auto s2 = locked(s1, work2, mutex): 
auto s3 = work3(s2);
sync_wait(s3);
```

That would be equivalent to:
``` 
auto sync_wait_scheduler = /* how to */;
auto stranded_scheduler = make_strand_scheduler(sync_wait_scheduler, mutex);
auto s1 = on(work1(), sync_wait_scheduler);
auto s2 = on(work2(s1), stranded_scheduler);
auto s3 = work3(s2);
sync_wait(s3);
```
[ Note: I need to mention the scheduler. ]

I believe something like this would be possible:
```
auto s1 = work1();
auto ls1 = lock(s1, mutex);
auto s2 = work2(ls1);
auto ls2 = unlock(s2);
auto s3 = work3(ls2);
sync_wait(s3);
```

But then I would have a separate `lock`/`unlock` operation. This is what I mean by RAII interface - I am passing `work2` as a callable, representing work that needs to be done under the mutex. But of course, my `locked` interface could be built on top of the `lock`/`unlock`.

**Comment by tomaszkam** [comment]
I have now realized why I am not using the `strand`/`schedule` based interface. One of my goals was to avoid any memory allocations, including one related to the serialization queue. Using the interface in the form:
`auto ls = locked(sender s, continuation c, async_mutex m)`
Where `s` is function that accepts any `sender` that has the same interface `s`, allows me to use `operation_states` as the nodes of serialization queue (they have `prev`/`next` pointer and virtual `run` method). The `sync_mutex` contains only `head`/`tail` pointers of the queue.

So, when `ls` is connected to the `receiver` `r` then,
1) connects `s` to the `lock_receiver` - the operation state contains a placeholder for nested `op_state` (see below)
2) fullfillment of `lock_receiver` connects `c(restore_via(args, r.get_scheduler))` with `unlock_receiver` and save the value in `op_state`, the `op_state` is linked into the `async_mutex` queue. If it is only element, the work is imediatelly started.
3) fulfillment of `unlock_receiver` removes the current node from the queue, start the next operation (assuming non-block), and then calls the appropriate method on `r`.

To achieve that, I need to pass the pointer to "locked" `op_state` between the `lock` and `unlock` operations.

**Comment by tomaszkam** [comment]
Secondary question, with the strand model, what happens if `foo` reschedules or forks (on the same scheduler) in:
`locked(foo, mutex)`.
Wouldn't the additional work be placed in serialization queue, potentially after another work? So some work can run in between. That would lead to potentially other work mutating the `data`.

With my model, the `locked(s, work, mutex)` we guarantee that transaction (not other work) is running between `s` calling `set_*` and `work(s)` calling `set_*`. This is regardless if the scheduler is switched or work is forked.
